FEXHeaderUtils/FEXHeaderUtils/Filesystem.h

```diff
@@ -14,4 +22,270 @@ namespace FHU::Filesystem {
+  }
+
+  inline bool IsRelative(const fextl::string &Path) {
+    return Path.starts_with('/');
+  }
+
+  enum class CopyOptions {
+    NONE,
+    SKIP_EXISTING,
+    OVERWRITE_EXISTING,
+  };
+
+  inline bool CopyFile(const fextl::string &From, const fextl::string &To, CopyOptions Options = CopyOptions::NONE) {
+    bool DestTested = false;
+    bool DestExists = false;
+    if (Options == CopyOptions::SKIP_EXISTING && Exists(To.c_str())) {
+      DestTested = true;
+      DestExists = Exists(To.c_str());
+      if (DestExists) {
+        // If the destination file exists already and the skip existing flag is set then
+        // return true without error.
+        return true;
+      }
+    }
+
+    if (!DestTested) {
+      DestTested = true;
+      DestExists = Exists(To.c_str());
+    }
+
+    if (Options == CopyOptions::OVERWRITE_EXISTING && DestExists) {
+      // If we are overwriting and the file exists then we want to use `sendfile` to overwrite
+      int SourceFD = open(From.c_str(), O_RDONLY | O_CLOEXEC);
+      if (SourceFD == -1) {
+        return false;
+      }
+
+      int DestinationFD = open(To.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0200);
+      if (DestinationFD == -1) {
+        close(SourceFD);
+        return false;
+      }
+
+      struct stat buf;
+      if (fstat(SourceFD, &buf) != 0) {
+        close(DestinationFD);
+        close(SourceFD);
+        return false;
+      }
+
+      // Set the destination permissions to the original source permissions.
+      if (fchmod(DestinationFD, buf.st_mode) != 0) {
+        close(DestinationFD);
+        close(SourceFD);
+        return false;
+      }
+      bool Result = sendfile(DestinationFD, SourceFD, nullptr, buf.st_size) == buf.st_size;
+      close(DestinationFD);
+      close(SourceFD);
+      return Result;
+    }
+
+    if (!DestExists) {
+      // If the destination doesn't exist then just use rename.
+      return rename(From.c_str(), To.c_str()) == 0;
+    }
+
+    return false;
+  }
+
+  inline fextl::string LexicallyNormal(const fextl::string &Path) {
+    const auto PathSize = Path.size();
+
+    // Early exit on empty paths.
+    if (PathSize == 0) {
+      return {};
+    }
+
+    // Count the number of separators up front
+    const auto SeparatorCount = std::count(Path.begin(), Path.end(), '/');
+
+    // Needs to be a list so iterators aren't invalidated while erasing.
+    // Maximum number of list regions will be the counted separators plus 2.
+    // Multiplied by two since it allocates * 2 the size (32-bytes per element)
+    // Use a small stack allocator to be more optimal.
+    // Needs to be a list so iterators aren't invalidated while erasing.
+    size_t DataSize = sizeof(std::string_view) * (SeparatorCount + 2) * 2;
+    void *Data = alloca(DataSize);
+    fextl::pmr::fixed_size_monotonic_buffer_resource mbr(Data, DataSize);
+    std::pmr::polymorphic_allocator pa {&mbr};
+    std::pmr::list<std::string_view> Parts{pa};
```

Comment:
Is per-element size of the `next` and `previous` pointers in the list considered in the DataSize above?

Reply:
Yes, that's what the comment above that you were confused about. Rewrote that part of the comment to explain more.